src/libsyncengine/update_detection/file_system_observer/fsoperationset.cpp

```diff
@@ -17,62 +17,57 @@
  */
 
 #include "fsoperationset.h"
+#include <cassert>
 
 namespace KDC {
 
 FSOperationSet::~FSOperationSet() {
     clear();
 }
 
-bool FSOperationSet::getOp(UniqueId id, FSOpPtr &opPtr, bool lockMutex /*= true*/) {
-    if (lockMutex) {
-        _mutex.lock();
-    }
+bool FSOperationSet::getOp(UniqueId id, FSOpPtr &opPtr) {
+    const std::scoped_lock lock(_mutex);
+    return getOpWithoutLock(id, opPtr);
+}
 
-    auto it = _ops.find(id);
-    if (it != _ops.end()) {
+bool FSOperationSet::getOpWithoutLock(UniqueId id, FSOpPtr &opPtr) {
+    assert(_mutex.try_lock() == false);
+    if (auto it = _ops.find(id); it != _ops.end()) {
         opPtr = it->second;
-        if (lockMutex) {
-            _mutex.unlock();
-        }
         return true;
     }
-
-    if (lockMutex) {
-        _mutex.unlock();
-    }
     return false;
 }
 
-bool FSOperationSet::getOpsByType(const OperationType type, std::unordered_set<UniqueId> &ops) {
-    const std::lock_guard<std::mutex> lock(_mutex);
-    auto it = _opsByType.find(type);
-    if (it != _opsByType.end()) {
+void FSOperationSet::getOpsByType(const OperationType type, std::unordered_set<UniqueId> &ops) {
+    const std::scoped_lock lock(_mutex);
+    if (auto it = _opsByType.find(type); it != _opsByType.end()) {
         ops = it->second;
-        return true;
+    } else {
+        ops.clear();
     }
-    return false;
 }
 
 bool FSOperationSet::getOpsByNodeId(const NodeId &nodeId, std::unordered_set<UniqueId> &ops) {
-    const std::lock_guard<std::mutex> lock(_mutex);
-    auto it = _opsByNodeId.find(nodeId);
-    if (it != _opsByNodeId.end()) {
+    const std::scoped_lock lock(_mutex);
+    if (auto it = _opsByNodeId.find(nodeId); it != _opsByNodeId.end()) {
         ops = it->second;
         return true;
     }
+    ops.clear();
```

Comment:
As `ops.clear()` is unnecessary when we find the nodeId (which occurs most of the time), I have excluded it from the if statement for optimization (to avoid an unnecessary call).

Reply:
The underlining question here is : is it the responsibility of the called function to initialize (clear in this context) the variable passed by reference? Or is it the responsibility of the caller?

Anyway, we could also never clear `ops` but return a boolean value, `true` if found, `false` if not. That way, the caller can discard the returned value if not found.

Reply:
> is it the responsibility of the called function to initialize (clear in this context) the variable passed by reference? Or is it the responsibility of the caller?

In my opinion, it is the responsibility of the called function to initialize (clear in this context) the variable passed by reference. This approach ensures that the result is consistent and reliable, as the called function has full control over the state of the variable. 
Additionally, it makes the code easier to maintain and debug, as there is only one place where the variable critticaly need to be initialized, rather than multiple places throughout the codebase.
(found an issue recently with UpdateInfo::parseString acting randomly because of this).

> Anyway, we could also never clear ops but return a boolean value, true if found, false if not. That way, the caller can discard the returned value if not found.

Most of the time, the result goes to a for loop, by assuring here that ops contains only the desired element discharge the caller for checking an other other boolean (when needed, it still can check the return value of the function wich correspond to found or not found).  It will just exit the loop immediatly. 
But it could maybe make the call function more explicit on how it behave?

> Actually, wouldn't it be simpler to let the function return a set of IDs?

It could be a solution yes. I think the time complexity would be the same as anyway we already do a copy

Reply:
> In my opinion, it is the responsibility of the called function to initialize (clear in this context) the variable passed by reference. This approach ensures that the result is consistent and reliable, as the called function has full control over the state of the variable.

Ok for me. As long as it is clear in the code. So IMHO we can keep it like that or apply Luc's suggestion